lib/common/patchset.c

```diff
@@ -316,61 +321,130 @@ xml_create_patchset_v2(xmlNode *source, xmlNode *target)
+
+        target_num_updates = 0;
+        crm_xml_add_int(target, XML_ATTR_NUMUPDATES, target_num_updates);
+
+    } else {
+        crm_xml_add_int(target, XML_ATTR_NUMUPDATES, ++target_num_updates);
+    }
+
+    crm_trace("%s changed (%d.%d.%d -> %d.%d.%d)",
+              (config_changed? "Config" : "Status"),
+              source_generation_admin, source_generation, source_num_updates,
+              target_generation_admin, target_generation, target_num_updates);
+}
+
+/*!
+ * \internal
+ * \brief Create an XML patchset showing CIB changes
+ *
+ * \param[in]     source           XML object before changes
+ * \param[in,out] target           XML object with changes applied
+ * \param[out]    config_changed   Where to store whether the configuration
+ *                                 changed
+ * \param[in]     manage_counters  Whether to update \c XML_ATTR_GENERATION or
+ *                                 \c XML_ATTR_NUMUPDATES counters in \p target
+ *                                 as appropriate
+ *
+ * \note This function always creates a v2 patchset. See the \c diff API schema
+ *       for details on format.
+ * \note If change tracking was not enabled while making changes, this function
+ *       returns \c NULL.
+ */
 xmlNode *
-xml_create_patchset(int format, xmlNode *source, xmlNode *target,
-                    bool *config_changed, bool manage_version)
+pcmk__xml_create_patchset(const xmlNode *source, xmlNode *target,
+                          bool *config_changed, bool manage_version)
 {
-    int counter = 0;
-    bool config = FALSE;
-    xmlNode *patch = NULL;
-    const char *version = crm_element_value(source, XML_ATTR_CRM_VERSION);
+    bool config = false;
+
+    CRM_CHECK((source != NULL) && (target != NULL), return NULL);
 
     xml_acl_disable(target);
+
     if (!xml_document_dirty(target)) {
-        crm_trace("No change %d", format);
-        return NULL; /* No change */
+        crm_trace("No change");
+        return NULL;
     }
 
     config = is_config_change(target);
-    if (config_changed) {
+    if (config_changed != NULL) {
         *config_changed = config;
     }
 
-    if (manage_version && config) {
-        crm_trace("Config changed %d", format);
-        crm_xml_add(target, XML_ATTR_NUMUPDATES, "0");
-
-        crm_element_value_int(target, XML_ATTR_GENERATION, &counter);
-        crm_xml_add_int(target, XML_ATTR_GENERATION, counter+1);
-
-    } else if (manage_version) {
-        crm_element_value_int(target, XML_ATTR_NUMUPDATES, &counter);
-        crm_trace("Status changed %d - %d %s", format, counter,
-                  crm_element_value(source, XML_ATTR_NUMUPDATES));
-        crm_xml_add_int(target, XML_ATTR_NUMUPDATES, (counter + 1));
+    if (manage_version) {
+        update_counters(source, target, config);
     }
+    return xml_create_patchset_v2(source, target);
+}
 
-    if (format == 0) {
-        if (compare_version("3.0.8", version) < 0) {
```

Comment:
.......That's embarrassing. I don't know what led me to thinking 3032878 created v2 diffs. The `CRM_FEATURE_SET` doc comment is very explicit about it starting with 3.0.9.

Maybe I git-blamed a different `compare_version()` call from the relevant one, IDK. I was very skeptical when I "found" this, since I figured the code would've already been dropped if possible, so I'm especially surprised I still went down this path.

Anyway, like you said, we can adapt this for 3.0.0 when the time comes.

---

Deprecating `cib_mixed_update` was the impetus for all this. In service of that, at one point I wanted to extend `cib__element_in_patchset()`. I needed to handle the v1 patchset case, and I didn't know how because the v1 patchset format is unclear.

I'll have to see if that's still necessary or if I have a clear path to deprecating `cib_mixed_update` anyway.